MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoInfo.java

```diff
@@ -0,0 +1,129 @@
+     */
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    private MotionPhotoInfo(MediaFormat mediaFormat, int videoOffset, long presentationTimestampUs) {
+        width = mediaFormat.getInteger(MediaFormat.KEY_WIDTH);
+        height = mediaFormat.getInteger(MediaFormat.KEY_HEIGHT);
+        duration = mediaFormat.getLong(MediaFormat.KEY_DURATION);
+        if (mediaFormat.containsKey(MediaFormat.KEY_ROTATION)) {
+            rotation = mediaFormat.getInteger(MediaFormat.KEY_ROTATION);
+        }
+        else {
+            rotation = 0;
+        }
+
+        this.videoOffset = videoOffset;
+        this.presentationTimestampUs = presentationTimestampUs;
+    }
+
+    /**
+     * Returns a new instance of MotionPhotoInfo for a specified file.
+     */
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    public static MotionPhotoInfo newInstance(String filename) throws IOException, XMPException {
+        return MotionPhotoInfo.newInstance(filename, new MediaExtractor());
+    }
+
+    /**
+     * Returns a new instance of MotionPhotoInfo with a specified file and MediaExtractor. Used for testing.
+     */
+    @VisibleForTesting
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    static MotionPhotoInfo newInstance(String filename, MediaExtractor extractor) throws IOException, XMPException {
+        XMPMeta meta = getFileXmp(filename);
+        int videoOffset = meta.getPropertyInteger("http://ns.google.com/photos/1.0/camera/", "MicroVideoOffset");
+        long presentationTimestampUs = meta.getPropertyLong("http://ns.google.com/photos/1.0/camera/", "MicroVideoPresentationTimestampUs");
+
+        MediaFormat mediaFormat = getFileMediaFormat(filename, extractor, videoOffset);
+        return new MotionPhotoInfo(mediaFormat, videoOffset, presentationTimestampUs);
+    }
+
+    /**
+     * Extract the JPEG XMP metadata from the Motion Photo.
+     */
+    private static XMPMeta getFileXmp(String filename) throws IOException, XMPException {
+        return XmpParser.getXmpMetadata(filename);
+    }
+
+    /**
+     * Get the MediaFormat associated with the video track of the Motion Photo MPEG4.
+     */
+    private static MediaFormat getFileMediaFormat(String filename, MediaExtractor extractor, int videoOffset) throws IOException {
+        File f = new File(filename);
+        try (FileInputStream fileInputStream = new FileInputStream(f)) {
+            FileDescriptor fd = fileInputStream.getFD();
+            extractor.setDataSource(fd, f.length() - videoOffset, videoOffset);
+            // Find the video track and create an appropriate media decoder
+            for (int i = 0; i < extractor.getTrackCount(); i++) {
+                MediaFormat format = extractor.getTrackFormat(i);
+                String mime = format.getString(MediaFormat.KEY_MIME);
+                assert mime != null;
+                if (mime.startsWith("video/")) {
+                    fileInputStream.close();
```

Comment:
this close call already gets called by the try (...) block. Delete the `fileInputStream.close();` line